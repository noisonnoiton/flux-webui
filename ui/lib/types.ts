export const AllNamespacesOption = "all";

export type NamespaceLabel = { value: string; label: string };
