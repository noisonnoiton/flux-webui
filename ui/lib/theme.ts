import green from "@material-ui/core/colors/green";
import purple from "@material-ui/core/colors/purple";
import { createMuiTheme } from "@material-ui/core/styles";
import { createGlobalStyle } from "styled-components";

const theme = createMuiTheme({
  typography: {
    fontFamily: "Montserrat",
  },
  palette: {
    primary: {
      main: purple[500],
    },
    secondary: {
      main: green[500],
    },
  },
});

export default theme;

export const GlobalStyle = createGlobalStyle`
  body {
    font-family: ${(props: { theme: typeof theme }) =>
      props.theme.typography.fontFamily}, sans-serif;
    padding: 0;
    margin: 0;
  }
`;
