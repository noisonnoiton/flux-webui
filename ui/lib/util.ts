import _ from "lodash";
import qs from "query-string";
import { toast } from "react-toastify";
import { DefaultClusters } from "./rpc/clusters";

export const wrappedFetch = (url, opts: RequestInit = {}) => {
  return fetch(url, {
    ...opts,
    credentials: "same-origin",
    headers: {
      "content-type": "application/json",
      ...(opts.headers || {}),
    },
  });
};

export const normalizePath = (pathname) => {
  return _.tail(pathname.split("/"));
};

export const prefixRoute = (route: string, ...idParams: string[]) =>
  `/:context/:namespace/${route}${
    idParams ? _.map(idParams, (p) => "/:" + p).join("") : ""
  }`;

export const toRoute = (route: PageRoute, params: string[]) => {
  const path = `/${_.map(params, (p) => `${p}/`).join("")}`;

  if (route === PageRoute.Home) {
    return route;
  }

  return `/${route}${params ? path : ""}`;
};

export const formatURL = (
  page: string,
  context: string,
  namespace: string,
  query: any = {}
) => {
  return `${page}?${qs.stringify({ context, namespace, ...query })}`;
};

export enum PageRoute {
  Redirector = "/",
  Home = "/sources",
  Sources = "/sources",
  SourceDetail = "/sources_detail",
  Kustomizations = "/kustomizations",
  KustomizationDetail = "/kustomizations_detail",
  HelmReleases = "/helmreleases",
  HelmReleaseDetail = "/helmreleases_detail",
  Events = "/events",
  Error = "/error",
}

export const getNavValue = (currentPage: any): PageRoute | boolean => {
  switch (currentPage) {
    case "kustomizations":
    case "kustomizations_detail":
      return PageRoute.Kustomizations;
    case "sources":
    case "sources_detail":
      return PageRoute.Sources;

    case "helmreleases":
    case "helmreleases_detail":
      return PageRoute.HelmReleases;

    case "events":
      return PageRoute.Events;

    default:
      return false;
  }
};

export const clustersClient = new DefaultClusters(
  "/api/clusters",
  wrappedFetch
);

export function notifySuccess(message: string) {
  toast["success"](message);
}

export function notifyError(message: string) {
  toast["error"](`Error: ${message}`);
}
