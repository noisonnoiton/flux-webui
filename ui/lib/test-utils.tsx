import { MuiThemeProvider } from "@material-ui/core";
import _ from "lodash";
import * as React from "react";
import { MemoryRouter } from "react-router";
import { ThemeProvider } from "styled-components";
import AppStateProvider from "../components/AppStateProvider";
import {
  GetChildObjectsRes,
  GetReconciledObjectsRes,
  Kustomization,
  ListContextsRes,
  ListKustomizationsRes,
  ListNamespacesForContextRes,
} from "./rpc/clusters";
import theme from "./theme";

type ClientOverrides = {
  listContexts: ListContextsRes;
  listNamespacesForContext: ListNamespacesForContextRes;
  listKustomizations?: ListKustomizationsRes;
  getReconciledObjects?: GetReconciledObjectsRes;
  getChildObjects?: GetChildObjectsRes;
};

export const createMockClient = (ovr: ClientOverrides) => {
  // Don't make the user wire up all the promise stuff to be interface-compliant
  const promisified = _.reduce(
    ovr,
    (result, desiredResponse, method) => {
      result[method] = () =>
        new Promise((accept) => accept(desiredResponse as any));

      return result;
    },
    {}
  );

  return promisified;
};

export function withContext(
  TestComponent,
  simulatedUrl: string,
  clientOverrides?: ClientOverrides
) {
  return (
    <MuiThemeProvider theme={theme}>
      <ThemeProvider theme={theme}>
        <MemoryRouter initialEntries={[simulatedUrl]}>
          <AppStateProvider clustersClient={createMockClient(clientOverrides)}>
            <TestComponent />
          </AppStateProvider>
        </MemoryRouter>
      </ThemeProvider>
    </MuiThemeProvider>
  );
}

const k: Kustomization = {
  name: "my-kustomization",
  namespace: "default",
  path: "/k8s",
  sourceref: "my-source",
  conditions: [],
  interval: "1m",
  prune: false,
  reconcilerequestat: "",
  reconcileat: "",
  sourcerefkind: "Git",
};

export const stubbedClustersResponses = {
  listContexts: {
    contexts: [{ name: "my-context" }, { name: "other-context" }],
    currentcontext: "other-context",
  },
  listNamespacesForContext: { namespaces: ["default"] },
  listKustomizations: {
    kustomizations: [k],
  },
  getReconciledObjects: {
    objects: [
      {
        groupversionkind: {
          group: "apps",
          version: "v1",
          kind: "Deployment",
        },
        name: "reconciled-deployment",
        namespace: "default",
        status: "Current",
      },
    ],
  },
  getChildObjects: {
    objects: [
      {
        groupversionkind: {
          group: "",
          version: "v1",
          kind: "Pod",
        },
        name: "reconciled-deployment-abc",
        namespace: "default",
        status: "Current",
      },
    ],
  },
};
