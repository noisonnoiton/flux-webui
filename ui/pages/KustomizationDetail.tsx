import {
  Box,
  Breadcrumbs,
  Button,
  CircularProgress,
  Tab,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
} from "@material-ui/core";
import { TabContext, TabList, TabPanel } from "@material-ui/lab";
import _ from "lodash";
import * as React from "react";
import styled from "styled-components";
import DataTable from "../components/DataTable";
import Flex from "../components/Flex";
import KeyValueTable from "../components/KeyValueTable";
import Link from "../components/Link";
import Page from "../components/Page";
import Panel from "../components/Panel";
import ReconciliationGraph from "../components/ReconciliationGraph";
import { useKubernetesContexts, useNavigation } from "../lib/hooks/app";
import {
  PARENT_CHILD_LOOKUP,
  UnstructuredObjectWithParent,
  useKustomizations,
} from "../lib/hooks/kustomizations";
import { Kustomization } from "../lib/rpc/clusters";
import { formatURL, PageRoute } from "../lib/util";

type Props = {
  className?: string;
};

const infoFields = [
  "sourceref",
  "namespace",
  "path",
  "interval",
  "prune",
  "lastappliedrevision",
];

const formatInfo = (detail: Kustomization) =>
  _.map(_.pick(detail, infoFields), (v, k) => ({
    key: k,
    value: typeof v === "string" ? v : v.toString(),
  }));

function KustomizationDetail({ className }: Props) {
  const [syncing, setSyncing] = React.useState(false);
  const [selectedTab, setSelectedTab] = React.useState("graph");
  const [reconciledObjects, setReconciledObjects] = React.useState<
    UnstructuredObjectWithParent[]
  >([]);
  const { query } = useNavigation();
  const { currentContext, currentNamespace } = useKubernetesContexts();

  const {
    kustomizations,
    syncKustomization,
    getReconciledObjects,
    getChildrenRecursive,
  } = useKustomizations(currentContext, currentNamespace);
  const kustomizationDetail = kustomizations[query.kustomizationId as string];

  const handleSyncClicked = () => {
    setSyncing(true);

    syncKustomization(kustomizationDetail).then(() => {
      setSyncing(false);
    });
  };

  React.useEffect(() => {
    if (!kustomizationDetail) {
      return;
    }

    const { snapshots, name, namespace } = kustomizationDetail;

    const kinds = _.reduce(
      snapshots,
      (r, e) => {
        r = [...r, ...e.kinds];
        return r;
      },
      []
    );

    const uniq = _.uniqBy(kinds, "kind");

    const getChildren = async () => {
      const { objects } = await getReconciledObjects(name, namespace, uniq);

      const result = [];
      for (let o = 0; o < objects.length; o++) {
        const obj = objects[o];

        await getChildrenRecursive(result, obj, PARENT_CHILD_LOOKUP);
      }

      setReconciledObjects(_.flatten(result));
    };

    let timeout;
    // function poll() {
    //   timeout = setTimeout(async () => {
    //     // Polling will stop if this errors.
    //     // Also prevents sending a request before the previous request finishes.
    //     await getChildren();
    //     poll();
    //   }, 5000);
    // }

    // Get children now, to avoid waiting for the first poll() setTimeout.
    getChildren();
    // Turn off polling for now.
    // The idea was to "live-update" the visualization,
    // but there are issues with the zoom causing weirdness with the
    // foreignObject html scaling when the graph re-renders.
    // poll();

    // Stop polling when the component unmounts
    return () => clearTimeout(timeout);
  }, [kustomizationDetail]);

  if (!kustomizationDetail) {
    return null;
  }

  const overrides = {
    sourceref: [
      <Link
        to={formatURL(
          PageRoute.SourceDetail,
          currentContext,
          currentNamespace,
          {
            sourceType: kustomizationDetail.sourcerefkind.toLowerCase(),
            sourceId: kustomizationDetail.sourceref,
          }
        )}
      >
        {kustomizationDetail.sourceref}
      </Link>,
      "Source",
    ],
    lastappliedrevision: [
      kustomizationDetail.lastappliedrevision,
      "Applied Revision",
    ],
  };

  return (
    <Page className={className}>
      <Flex align wide>
        <Breadcrumbs>
          <Link
            to={formatURL(
              PageRoute.Kustomizations,
              currentContext,
              currentNamespace
            )}
          >
            <h2>Kustomizations</h2>
          </Link>
          <Flex wide>
            <h2>{kustomizationDetail.name}</h2>
          </Flex>
        </Breadcrumbs>
        <Button
          onClick={handleSyncClicked}
          color="primary"
          disabled={syncing}
          variant="contained"
        >
          {syncing ? <CircularProgress size={24} /> : "Sync"}
        </Button>
      </Flex>

      <Box marginBottom={2}>
        <Panel title="Info">
          <KeyValueTable
            columns={3}
            pairs={formatInfo(kustomizationDetail)}
            overrides={overrides}
          />
        </Panel>
      </Box>

      <Box marginBottom={2}>
        <Panel title="Conditions">
          <TableContainer>
            <Table>
              <TableHead>
                <TableRow>
                  <TableCell>Type</TableCell>
                  <TableCell>Status</TableCell>
                  <TableCell>Timestamp</TableCell>
                  <TableCell>Reason</TableCell>
                  <TableCell>Message</TableCell>
                </TableRow>
              </TableHead>
              <TableBody>
                {_.map(kustomizationDetail.conditions, (c, i) => (
                  <TableRow key={i}>
                    <TableCell>{c.type}</TableCell>
                    <TableCell>{c.status}</TableCell>
                    <TableCell>
                      {new Date(c.timestamp).toLocaleDateString()}
                    </TableCell>
                    <TableCell>{c.reason}</TableCell>
                    <TableCell>{c.message}</TableCell>
                  </TableRow>
                ))}
              </TableBody>
            </Table>
          </TableContainer>
        </Panel>
      </Box>
      <Box marginBottom={2}>
        <Panel title="Reconciled Objects">
          <TabContext value={selectedTab}>
            <TabList onChange={(e, i) => setSelectedTab(i)}>
              <Tab id="graph-button" label="Graph" value="graph" />
              <Tab id="table-button" label="Table" value="table" />
            </TabList>
            <TabPanel value="graph">
              <ReconciliationGraph
                objects={reconciledObjects}
                parentObject={kustomizationDetail}
                parentObjectKind="Kustomization"
              />
            </TabPanel>
            <TabPanel value="table">
              <DataTable
                sortFields={["name", "kind"]}
                fields={[
                  { label: "Kind", value: (v) => v.groupversionkind.kind },
                  { label: "Name", value: "name" },
                  { label: "Namespace", value: "namespace" },
                  { label: "Status", value: "status" },
                ]}
                rows={reconciledObjects}
              />
            </TabPanel>
          </TabContext>
        </Panel>
      </Box>
    </Page>
  );
}

export default styled(KustomizationDetail)`
  .MuiBreadcrumbs-root {
    width: 100%;
  }
`;
