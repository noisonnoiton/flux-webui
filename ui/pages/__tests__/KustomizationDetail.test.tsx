import { render, screen } from "@testing-library/react";
import { stubbedClustersResponses, withContext } from "../../lib/test-utils";
import KustomizationDetail from "../KustomizationDetail";

// https://github.com/mui-org/material-ui/issues/21293
// MUI relies on Math.random() to generate IDs.
// Mock it to create deterministic tests.
const mockMath = Object.create(global.Math);
mockMath.random = () => 0.5;
global.Math = mockMath;

describe("KustomizationDetail", () => {
  let container;
  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });
  afterEach(() => {
    document.body.removeChild(container);
    container = null;
  });
  it("renders", async () => {
    const url =
      "/?context=my-context&namespace=default&kustomizationId=my-kustomization";
    const tree = render(
      withContext(KustomizationDetail, url, stubbedClustersResponses),
      container
    );

    expect(
      (await screen.findByText("my-kustomization")).textContent
    ).toBeTruthy();
    expect(tree).toMatchSnapshot();
  });
});
