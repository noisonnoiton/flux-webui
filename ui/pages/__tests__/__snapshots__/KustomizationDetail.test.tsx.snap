// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KustomizationDetail renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div>
      <div
        class="sc-iBPRYJ sc-iqHYGH rMTDD"
      >
        <div
          class="sc-dlfnbm emcvKi"
        >
          <nav
            class="MuiTypography-root MuiBreadcrumbs-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            <ol
              class="MuiBreadcrumbs-ol"
            >
              <li
                class="MuiBreadcrumbs-li"
              >
                <a
                  class="sc-gKsewC"
                  href="/kustomizations?context=my-context&namespace=default"
                >
                  <h2>
                    Kustomizations
                  </h2>
                </a>
              </li>
              <li
                aria-hidden="true"
                class="MuiBreadcrumbs-separator"
              >
                /
              </li>
              <li
                class="MuiBreadcrumbs-li"
              >
                <div
                  class="sc-dlfnbm eIFobp"
                >
                  <h2>
                    my-kustomization
                  </h2>
                </div>
              </li>
            </ol>
          </nav>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Sync
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root MuiBox-root-1"
        >
          <div
            class="sc-pFZIQ"
          >
            <div
              class="MuiPaper-root MuiCard-root MuiPaper-outlined MuiPaper-rounded"
            >
              <div
                class="sc-fubCfw eQvqYT"
              >
                <h3>
                  Info
                </h3>
              </div>
              <div
                class="MuiCardContent-root"
              >
                <div>
                  <div
                    class="sc-jSgupP crqFgI"
                    role="list"
                  >
                    <table>
                      <tbody>
                        <tr>
                          <td
                            role="listitem"
                          >
                            <div
                              aria-label="Source"
                              class="sc-hKgILt ctfbPn"
                            >
                              Source
                            </div>
                            <div
                              class="sc-eCssSg"
                            >
                              <a
                                class="sc-gKsewC"
                                href="/sources_detail?context=my-context&namespace=default&sourceId=my-source&sourceType=git"
                              >
                                my-source
                              </a>
                            </div>
                          </td>
                          <td
                            role="listitem"
                          >
                            <div
                              aria-label="Namespace"
                              class="sc-hKgILt ctfbPn"
                            >
                              Namespace
                            </div>
                            <div
                              class="sc-eCssSg"
                            >
                              default
                            </div>
                          </td>
                          <td
                            role="listitem"
                          >
                            <div
                              aria-label="Path"
                              class="sc-hKgILt ctfbPn"
                            >
                              Path
                            </div>
                            <div
                              class="sc-eCssSg"
                            >
                              /k8s
                            </div>
                          </td>
                        </tr>
                        <tr>
                          <td
                            role="listitem"
                          >
                            <div
                              aria-label="Interval"
                              class="sc-hKgILt ctfbPn"
                            >
                              Interval
                            </div>
                            <div
                              class="sc-eCssSg"
                            >
                              1m
                            </div>
                          </td>
                          <td
                            role="listitem"
                          >
                            <div
                              aria-label="Prune"
                              class="sc-hKgILt ctfbPn"
                            >
                              Prune
                            </div>
                            <div
                              class="sc-eCssSg"
                            >
                              false
                            </div>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-2"
        >
          <div
            class="sc-pFZIQ"
          >
            <div
              class="MuiPaper-root MuiCard-root MuiPaper-outlined MuiPaper-rounded"
            >
              <div
                class="sc-fubCfw eQvqYT"
              >
                <h3>
                  Conditions
                </h3>
              </div>
              <div
                class="MuiCardContent-root"
              >
                <div>
                  <div
                    class="MuiTableContainer-root"
                  >
                    <table
                      class="MuiTable-root"
                    >
                      <thead
                        class="MuiTableHead-root"
                      >
                        <tr
                          class="MuiTableRow-root MuiTableRow-head"
                        >
                          <th
                            class="MuiTableCell-root MuiTableCell-head"
                            scope="col"
                          >
                            Type
                          </th>
                          <th
                            class="MuiTableCell-root MuiTableCell-head"
                            scope="col"
                          >
                            Status
                          </th>
                          <th
                            class="MuiTableCell-root MuiTableCell-head"
                            scope="col"
                          >
                            Timestamp
                          </th>
                          <th
                            class="MuiTableCell-root MuiTableCell-head"
                            scope="col"
                          >
                            Reason
                          </th>
                          <th
                            class="MuiTableCell-root MuiTableCell-head"
                            scope="col"
                          >
                            Message
                          </th>
                        </tr>
                      </thead>
                      <tbody
                        class="MuiTableBody-root"
                      />
                    </table>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-3"
        >
          <div
            class="sc-pFZIQ"
          >
            <div
              class="MuiPaper-root MuiCard-root MuiPaper-outlined MuiPaper-rounded"
            >
              <div
                class="sc-fubCfw eQvqYT"
              >
                <h3>
                  Reconciled Objects
                </h3>
              </div>
              <div
                class="MuiCardContent-root"
              >
                <div>
                  <div
                    class="MuiTabs-root"
                  >
                    <div
                      class="MuiTabs-scroller MuiTabs-fixed"
                      style="overflow: hidden;"
                    >
                      <div
                        class="MuiTabs-flexContainer"
                        role="tablist"
                      >
                        <button
                          aria-controls="mui-p-50000-P-graph"
                          aria-selected="true"
                          class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit Mui-selected"
                          id="mui-p-50000-T-graph"
                          role="tab"
                          tabindex="0"
                          type="button"
                        >
                          <span
                            class="MuiTab-wrapper"
                          >
                            Graph
                          </span>
                          <span
                            class="MuiTouchRipple-root"
                          />
                        </button>
                        <button
                          aria-controls="mui-p-50000-P-table"
                          aria-selected="false"
                          class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
                          id="mui-p-50000-T-table"
                          role="tab"
                          tabindex="-1"
                          type="button"
                        >
                          <span
                            class="MuiTab-wrapper"
                          >
                            Table
                          </span>
                          <span
                            class="MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                      <span
                        class="PrivateTabIndicator-root-4 PrivateTabIndicator-colorSecondary-6 MuiTabs-indicator"
                        style="left: 0px; width: 0px;"
                      />
                    </div>
                  </div>
                  <div
                    aria-labelledby="mui-p-50000-T-graph"
                    class="MuiTabPanel-root"
                    id="mui-p-50000-P-graph"
                    role="tabpanel"
                  >
                    <div
                      class="sc-kEjbxe fvZZpi"
                    >
                      <div
                        class="sc-jrAGrp dTbqnH"
                      >
                        <svg
                          height="960"
                          width="960"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    aria-labelledby="mui-p-50000-T-table"
                    class="MuiTabPanel-root"
                    hidden=""
                    id="mui-p-50000-P-table"
                    role="tabpanel"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-iBPRYJ sc-iqHYGH rMTDD"
    >
      <div
        class="sc-dlfnbm emcvKi"
      >
        <nav
          class="MuiTypography-root MuiBreadcrumbs-root MuiTypography-body1 MuiTypography-colorTextSecondary"
        >
          <ol
            class="MuiBreadcrumbs-ol"
          >
            <li
              class="MuiBreadcrumbs-li"
            >
              <a
                class="sc-gKsewC"
                href="/kustomizations?context=my-context&namespace=default"
              >
                <h2>
                  Kustomizations
                </h2>
              </a>
            </li>
            <li
              aria-hidden="true"
              class="MuiBreadcrumbs-separator"
            >
              /
            </li>
            <li
              class="MuiBreadcrumbs-li"
            >
              <div
                class="sc-dlfnbm eIFobp"
              >
                <h2>
                  my-kustomization
                </h2>
              </div>
            </li>
          </ol>
        </nav>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Sync
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiBox-root MuiBox-root-1"
      >
        <div
          class="sc-pFZIQ"
        >
          <div
            class="MuiPaper-root MuiCard-root MuiPaper-outlined MuiPaper-rounded"
          >
            <div
              class="sc-fubCfw eQvqYT"
            >
              <h3>
                Info
              </h3>
            </div>
            <div
              class="MuiCardContent-root"
            >
              <div>
                <div
                  class="sc-jSgupP crqFgI"
                  role="list"
                >
                  <table>
                    <tbody>
                      <tr>
                        <td
                          role="listitem"
                        >
                          <div
                            aria-label="Source"
                            class="sc-hKgILt ctfbPn"
                          >
                            Source
                          </div>
                          <div
                            class="sc-eCssSg"
                          >
                            <a
                              class="sc-gKsewC"
                              href="/sources_detail?context=my-context&namespace=default&sourceId=my-source&sourceType=git"
                            >
                              my-source
                            </a>
                          </div>
                        </td>
                        <td
                          role="listitem"
                        >
                          <div
                            aria-label="Namespace"
                            class="sc-hKgILt ctfbPn"
                          >
                            Namespace
                          </div>
                          <div
                            class="sc-eCssSg"
                          >
                            default
                          </div>
                        </td>
                        <td
                          role="listitem"
                        >
                          <div
                            aria-label="Path"
                            class="sc-hKgILt ctfbPn"
                          >
                            Path
                          </div>
                          <div
                            class="sc-eCssSg"
                          >
                            /k8s
                          </div>
                        </td>
                      </tr>
                      <tr>
                        <td
                          role="listitem"
                        >
                          <div
                            aria-label="Interval"
                            class="sc-hKgILt ctfbPn"
                          >
                            Interval
                          </div>
                          <div
                            class="sc-eCssSg"
                          >
                            1m
                          </div>
                        </td>
                        <td
                          role="listitem"
                        >
                          <div
                            aria-label="Prune"
                            class="sc-hKgILt ctfbPn"
                          >
                            Prune
                          </div>
                          <div
                            class="sc-eCssSg"
                          >
                            false
                          </div>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root MuiBox-root-2"
      >
        <div
          class="sc-pFZIQ"
        >
          <div
            class="MuiPaper-root MuiCard-root MuiPaper-outlined MuiPaper-rounded"
          >
            <div
              class="sc-fubCfw eQvqYT"
            >
              <h3>
                Conditions
              </h3>
            </div>
            <div
              class="MuiCardContent-root"
            >
              <div>
                <div
                  class="MuiTableContainer-root"
                >
                  <table
                    class="MuiTable-root"
                  >
                    <thead
                      class="MuiTableHead-root"
                    >
                      <tr
                        class="MuiTableRow-root MuiTableRow-head"
                      >
                        <th
                          class="MuiTableCell-root MuiTableCell-head"
                          scope="col"
                        >
                          Type
                        </th>
                        <th
                          class="MuiTableCell-root MuiTableCell-head"
                          scope="col"
                        >
                          Status
                        </th>
                        <th
                          class="MuiTableCell-root MuiTableCell-head"
                          scope="col"
                        >
                          Timestamp
                        </th>
                        <th
                          class="MuiTableCell-root MuiTableCell-head"
                          scope="col"
                        >
                          Reason
                        </th>
                        <th
                          class="MuiTableCell-root MuiTableCell-head"
                          scope="col"
                        >
                          Message
                        </th>
                      </tr>
                    </thead>
                    <tbody
                      class="MuiTableBody-root"
                    />
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root MuiBox-root-3"
      >
        <div
          class="sc-pFZIQ"
        >
          <div
            class="MuiPaper-root MuiCard-root MuiPaper-outlined MuiPaper-rounded"
          >
            <div
              class="sc-fubCfw eQvqYT"
            >
              <h3>
                Reconciled Objects
              </h3>
            </div>
            <div
              class="MuiCardContent-root"
            >
              <div>
                <div
                  class="MuiTabs-root"
                >
                  <div
                    class="MuiTabs-scroller MuiTabs-fixed"
                    style="overflow: hidden;"
                  >
                    <div
                      class="MuiTabs-flexContainer"
                      role="tablist"
                    >
                      <button
                        aria-controls="mui-p-50000-P-graph"
                        aria-selected="true"
                        class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit Mui-selected"
                        id="mui-p-50000-T-graph"
                        role="tab"
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="MuiTab-wrapper"
                        >
                          Graph
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                      <button
                        aria-controls="mui-p-50000-P-table"
                        aria-selected="false"
                        class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
                        id="mui-p-50000-T-table"
                        role="tab"
                        tabindex="-1"
                        type="button"
                      >
                        <span
                          class="MuiTab-wrapper"
                        >
                          Table
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                    <span
                      class="PrivateTabIndicator-root-4 PrivateTabIndicator-colorSecondary-6 MuiTabs-indicator"
                      style="left: 0px; width: 0px;"
                    />
                  </div>
                </div>
                <div
                  aria-labelledby="mui-p-50000-T-graph"
                  class="MuiTabPanel-root"
                  id="mui-p-50000-P-graph"
                  role="tabpanel"
                >
                  <div
                    class="sc-kEjbxe fvZZpi"
                  >
                    <div
                      class="sc-jrAGrp dTbqnH"
                    >
                      <svg
                        height="960"
                        width="960"
                      />
                    </div>
                  </div>
                </div>
                <div
                  aria-labelledby="mui-p-50000-T-table"
                  class="MuiTabPanel-root"
                  hidden=""
                  id="mui-p-50000-P-table"
                  role="tabpanel"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
